File: Python3/3258_count_substrings_that_satisfy_k_constraint_i.py
# https://leetcode.cn/problems/count-substrings-that-satisfy-k-constraint-i/

class Solution:
    def countKConstraintSubstrings(self, s: str, k: int) -> int:
        res = 0
        i = 0
        cnt = [0, 0]
        for j, ch in enumerate(s):
            cnt[int(ch)] += 1
            while cnt[0] > k and cnt[1] > k and i <= j:
                cnt[int(s[i])] -= 1
                i += 1
            res += j - i + 1
        return res

File: Cpp/3258_count_substrings_that_satisfy_k_constraint_i.cpp
// https://leetcode.cn/problems/count-substrings-that-satisfy-k-constraint-i/

class Solution {
public:
    int countKConstraintSubstrings(string s, int k) {
        int res = 0;
        int i = 0, n = s.size();
        vector<int> cnt(2, 0);
        for (int j = 0; j < n; ++j) {
            char ch = s[j];
            cnt[ch - '0'] += 1;
            while (cnt[0] > k && cnt[1] > k && i <= j) {
                cnt[s[i] - '0'] -= 1;
                i += 1;
            }
            res += j - i + 1;
        }
        return res;
    }
};
